services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/UploadDirectoryOverrideConfiguration.java

```diff
@@ -156,6 +173,53 @@ public static Builder builder() {
          */
         Builder maxDepth(Integer maxDepth);
 
+        /**
+         * Specify a function used to transform the {@link UploadFileRequest}s generated by this {@link UploadDirectoryRequest}.
+         * The provided function is called once for each file that is uploaded, allowing you to modify the paths resolved by
+         * TransferManager on a per-file basis, modify the created {@link PutObjectRequest} before it is passed to S3, or
+         * configure a {@link TransferRequestOverrideConfiguration}.
+         *
+         * <p>The factory receives the {@link UploadFileRequest}s created by Transfer Manager for each file in the directory
+         * being uploaded, and returns a (potentially modified) {@code UploadFileRequest}.
+         *
+         * <p>
+         * <b>Usage Example:</b>
+         * <pre>
+         * {@code
+         * // Add a LoggingTransferListener to every transfer within the upload directory request
+         * TransferRequestOverrideConfiguration fileUploadConfiguration =
+         *     TransferRequestOverrideConfiguration.builder()
+         *         .addListener(LoggingTransferListener.create())
```

Comment:
Out of scope of this PR but worth noting: This works, but it's going to result in using the same listener for every invocation, when the current `LoggingTransferListener` implementation assumes that it's not going to be reused. We've so far refrained from adding listeners at the client-level for this reason. A client could implement a reusable listener by internally mapping something like `TransferRequest` to internal listener state, but this is a lot of complexity. I wonder if we should change listeners to be declared with a `Supplier` instead.

Reply:
I see, yeah, seems like it's easy to make that mistake, then. I certainly thought this would be fine. What do you recommend I do for the purposes of this PR?

Reply:
If we moved the `LoggingTransferListener.create()` call to be inside your lambda, it would create a new one for every request and work as expected, but that may make the code less readable. If we can update the example to do that, without severely affecting legibility, then I would recommend that. Otherwise I think it's okay to use your example as-is and we can do a fast-follow-up on whether to accept a `Supplier` or not.

Reply:
It would unfortuantely be very difficult to make it readable with the lambdas. I'll leave it as-is and we can move it to a supplier asynchronously.